### PR TITLE
[full-ci] Fix sharesTree loading

### PR DESCRIPTION
**Pull request by JammingBen**
## Description
We've improved loading of the shares tree:

* It now happens more globally in the sidebar component instead of in each sidebar panel.
* Shares won't be loaded for resources without a path anymore.

These changes massively improve the sidebar performance and fix several issues with (re-)share permissions.

Note: Needs ocis with the latest Reva update to include the `shareId`.

## Related Issue

- Fixes https://github.com/owncloud/web/issues/7506
- Fixes https://github.com/owncloud/web/issues/7593
- Fixes https://github.com/owncloud/web/issues/7592

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests

### CODE REVIEW

**Comment by JammingBen** [comment]
Shares loading via empty string `''` leads to ALL shares being loaded, which is a bad thing when loading shares for your personal root for example (and in general IMO). The only reason for this is to retrieve the parent share in the shares jail, so the parent permissions won't be exceeded.

Maybe we shouldn't go the shares tree route here. Instead we could retrieve the parent share via the shareId, which we got in the resource now. The only issue here is that we somehow need to store the parent share somewhere? Or always load it when the shares panel gets opened. @kulmann What do you think?

File: packages/web-app-files/src/components/SideBar/Shares/Collaborators/RoleDropdown.vue [hunk 1]
```diff
@@ -183,7 +183,7 @@ export default defineComponent({
     },
     share() {
       // the root share has an empty key in the shares tree. That's the reason why we retrieve the share by an empty key here
```
Comment (kulmann) [inline packages/web-app-files/src/components/SideBar/Shares/Collaborators/RoleDropdown.vue:185]: please adjust comment as well

File: packages/web-app-files/src/components/SideBar/Shares/FileLinks.vue [hunk 2]
```diff
@@ -184,7 +184,7 @@ export default defineComponent({
 
     share() {
       // the root share has an empty key in the shares tree. That's the reason why we retrieve the share by an empty key here
```
Comment (kulmann) [inline packages/web-app-files/src/components/SideBar/Shares/FileLinks.vue:186]: please adjust comment as well

**Comment by ownclouders** [comment]
Results for <strong>oC10SharingExternalRoot</strong> https://drone.owncloud.com/owncloud/web/28114/43/1
:boom: The oC10SharingExternalRoot tests pipeline failed. The build has been cancelled.

File: packages/web-app-files/src/helpers/path.js [hunk 3]
```diff
@@ -27,6 +27,10 @@ export function getParentPaths(path = '', includeCurrent = false) {
 
   sections.pop()
   while (sections.length > 0) {
+    if (!sections.join('/')) {
```
Comment (kulmann) [inline packages/web-app-files/src/helpers/path.js:30]: What about changing line 22 to
```js
const sections = s.split('/').filter(Boolean)
```
instead to remove empty strings from the array?

File: packages/web-app-files/src/components/SideBar/Shares/Collaborators/RoleDropdown.vue [hunk 4]
```diff
@@ -223,7 +223,8 @@ export default defineComponent({
     window.removeEventListener('keydown', this.cycleRoles)
   },
 
-  mounted() {
+  async mounted() {
+    await this.loadIncomingParentShare.perform(this.resource)
```
Comment (kulmann) [inline packages/web-app-files/src/components/SideBar/Shares/Collaborators/RoleDropdown.vue:227]: doesn't this try to load the incoming parent shares for each single `RoleDropdown` component instance? Can we load the incoming parent shares once in the `SideBar.vue` or in the `SharesPanel.vue` and hand it down to the child components via provide/inject (if from SideBar.vue) or via component properties (if coming from SharesPanel.vue)?

**Review by kulmann — CHANGES_REQUESTED** [review]

File: packages/web-app-files/src/helpers/path.js (hunk 3 above)
Comment (JammingBen) [reply packages/web-app-files/src/helpers/path.js:30]: That would actually result in a different array at the end. When `''` is removed beforehand, the leading slash would be missing.

File: packages/web-app-files/src/components/SideBar/Shares/Collaborators/ListItem.vue [hunk 5]
```diff
@@ -258,11 +251,6 @@ export default defineComponent({
       return this.sharedParentRoute?.params?.item.split('/').pop()
     },
 
-    shareDetailsHelperContent() {
```
Comment (JammingBen) [inline packages/web-app-files/src/components/SideBar/Shares/Collaborators/ListItem.vue:261]: Fyi unused property.

File: packages/web-app-files/src/components/SideBar/Shares/FileShares.vue [hunk 6]
```diff
@@ -188,38 +188,6 @@ export default {
       return this.collaborators.length > 0
     },
 
-    /**
```
Comment (JammingBen) [inline packages/web-app-files/src/components/SideBar/Shares/FileShares.vue:191]: Fyi unused property.

File: packages/web-app-files/src/components/SideBar/Shares/FileShares.vue [hunk 7]
```diff
@@ -294,20 +259,6 @@ export default {
       const translatedFolder = this.$gettext("You don't have permission to share this folder.")
       return this.highlightedFile.type === 'file' ? translatedFile : translatedFolder
     },
-
```
Comment (JammingBen) [inline packages/web-app-files/src/components/SideBar/Shares/FileShares.vue:297]: Fyi unused property.

File: packages/web-app-files/src/composables/parentShare/useIncomingParentShare.ts [hunk 8]
```diff
@@ -0,0 +1,34 @@
+import { buildShare } from '../../helpers/resources'
+import { useStore } from 'web-pkg/src/composables'
+import { computed, ref, unref } from '@vue/composition-api'
+import { useTask } from 'vue-concurrency'
+import { clientService } from 'web-pkg/src/services'
+
+export function useIncomingParentShare() {
+  const store = useStore()
+  const incomingParentShare = ref(null)
+  const sharesTree = computed(() => store.state.Files.sharesTree)
```
Comment (kulmann) [inline packages/web-app-files/src/composables/parentShare/useIncomingParentShare.ts:10]: Please use the getter for the sharesTree: `store.getters['Files/sharesTree']`.
Reason: In my experience that's a good idea for building backwards compatibility. If something from the state gets moved or removed you can write some logic in the getter to keep it compatible for a while and deprecate it. With state that's not possible. That being said I'd even recommend to never access vuex state directly (outside of the store module itself).

File: packages/web-app-files/src/composables/parentShare/useIncomingParentShare.ts [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+  const loadIncomingParentShare = useTask(function* (signal, resource) {
+    let parentShare
+    for (const shares of Object.values(unref(sharesTree)) as any) {
```
Comment (kulmann) [inline packages/web-app-files/src/composables/parentShare/useIncomingParentShare.ts:14]: Please walk the sharesTree from the leaf up to the root in sorted manner to get the closest match in case of parent shares on multiple levels.

Comment (kulmann) [reply packages/web-app-files/src/composables/parentShare/useIncomingParentShare.ts:14]: Can we ensure that the `sharesTree` is already loaded?

File: packages/web-app-files/src/composables/parentShare/useIncomingParentShare.ts [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+      parentShare = shares.find((s) => s.incoming)
+      if (parentShare) {
+        incomingParentShare.value = parentShare
+        return
+      }
+    }
+
+    if (resource.shareId) {
+      const parentShare = yield clientService.owncloudSdk.shares.getShare(resource.shareId)
```
Comment (kulmann) [inline packages/web-app-files/src/composables/parentShare/useIncomingParentShare.ts:23]: the `parentShare` variable is already defined above, so you should drop the `const` here, no?

**Review by kulmann — CHANGES_REQUESTED** [review]

File: packages/web-app-files/src/components/SideBar/Details/FileDetails.vue [hunk 11]
```diff
@@ -352,39 +352,41 @@ export default defineComponent({
...
 
-      this.sharedItem = shares[0]
-      this.sharedByName = this.sharedItem.owner?.name
-      this.sharedByDisplayName = this.sharedItem.owner?.displayName
-      if (this.sharedItem.owner?.additionalInfo) {
-        this.sharedByDisplayName += ' (' + this.sharedItem.owner.additionalInfo + ')'
-      }
-      this.sharedTime = this.sharedItem.stime
-      this.sharedParentDir = sharePathParentOrCurrent
+        this.sharedItem = shares[0]
```
[the remark this answers is not in the document]
Comment (JammingBen) [reply packages/web-app-files/src/components/SideBar/Details/FileDetails.vue:375]: Sounds worth trying, although I would like to do that in a separate PR.

File: packages/web-app-files/src/composables/parentShare/useIncomingParentShare.ts (hunk 9 above)
Comment (JammingBen) [reply packages/web-app-files/src/composables/parentShare/useIncomingParentShare.ts:14]: > Please walk the sharesTree from the leaf up to the root in sorted manner to get the closest match in case of parent shares on multiple levels.

It seems that this is already the case. Honestly, I don't know why, maybe the combination of `Object.values()` and `unref()` leads to this?

> Can we ensure that the sharesTree is already loaded?

Yes. `loadIncomingParentShare()` is being called in a `sharesTree` watcher in the `SharesPanel` component (= after the `sharesTree` has been loaded.

File: tests/e2e/support/objects/app-files/share/actions.ts [hunk 12]
```diff
@@ -81,13 +81,14 @@ export const inviteMembers = async (args: inviteMembersArgs): Promise<void> => {
...
     const shareInputLocator = page.locator(invitationInput)
+    await shareInputLocator.click()
     await Promise.all([
       page.waitForResponse((resp) => resp.url().includes('sharees') && resp.status() === 200),
       shareInputLocator.fill(recipient.id)
     ])
     await shareInputLocator.focus()
     await page.waitForSelector('.vs--open')
-    await page.locator(invitationInput).press('Enter')
+    await page.locator('.vs__dropdown-option').click()
```
Comment (JammingBen) [inline tests/e2e/support/objects/app-files/share/actions.ts:91]: I don't know why these changes are necessary, but without them, e2e tests won't succeed. I assume this happens because the panel is now loading way faster than before... but that's just a guess.

File: changelog/unreleased/bugfix-shares-tree-loading [hunk 13]
```diff
@@ -0,0 +1,12 @@
...
+We've improved loading of the shares tree:
+
+* It now happens more globally in the sidebar component instead of in each sidebar panel.
+* Shares won't be loaded for resources without a path anymore.
+
+These changes massively improve the sidebar performance and fix several issues with (re-)share permissions.
+
+https://github.com/owncloud/web/issues/7506
+https://github.com/owncloud/web/issues/7593
+https://github.com/owncloud/web/pull/7580
```
Comment (kulmann) [inline changelog/unreleased/bugfix-shares-tree-loading:12]: I added https://github.com/owncloud/web/issues/7592 to the PR description. Could you add it to the changelog as well?

File: packages/web-app-files/src/components/SideBar/SideBar.vue [hunk 14]
```diff
@@ -53,6 +53,8 @@ import {
 } from 'web-pkg/src/composables'
 import { bus } from 'web-pkg/src/instance'
 import { SideBarEventTopics } from '../../composables/sideBar'
+import _ from 'lodash'
```
Comment (kulmann) [inline packages/web-app-files/src/components/SideBar/SideBar.vue:56]: Please only import lodash functions via `lodash-es/<functionName>`

File: packages/web-app-files/src/components/SideBar/SideBar.vue [hunk 15]
```diff
@@ -197,6 +201,10 @@ export default defineComponent({
         return
       }
 
+      if (oldFile && _.isEqual(newFile, oldFile)) {
```
Comment (kulmann) [inline packages/web-app-files/src/components/SideBar/SideBar.vue:204]: Please import `lodash-es/isEqual` instead (see comment on import)

File: packages/web-app-files/src/composables/parentShare/useIncomingParentShare.ts (hunk 9 above)
Comment (kulmann) [reply packages/web-app-files/src/composables/parentShare/useIncomingParentShare.ts:14]: 👍

**Review by kulmann — CHANGES_REQUESTED** [review]
One last tiny nitpick (about lodash imports, see comment). Otherwise superb, love it!

**Review by kulmann — APPROVED** [review]
LGTM now 🚀
